File: one-string-swap/solution1790.go
// https://leetcode.com/problems/check-if-one-string-swap-can-make-strings-equal/

package main

import "fmt"

// T: O(n), S: O(1)
func areAlmostEqual(s1 string, s2 string) bool {
	counts1 := make([]int, 26)
	counts2 := make([]int, 26)
	diffs := 0
	for i := 0; i < len(s1); i++ {
		if s1[i] != s2[i] {
			diffs++
		}
		counts1[s1[i]-'a']++
		counts2[s2[i]-'a']++
	}
	if !arraysEqual(counts1, counts2) {
		return false
	}
	return diffs == 2 || diffs == 0
}

func arraysEqual(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(areAlmostEqual("bank", "kanb"))     // true
	fmt.Println(areAlmostEqual("attack", "defend")) // false
	fmt.Println(areAlmostEqual("kelb", "kelb"))     // true
	fmt.Println(areAlmostEqual("caa", "aaz"))       // false
}

File: one-string-swap/Solution1790.java
// https://leetcode.com/problems/check-if-one-string-swap-can-make-strings-equal/

import java.util.Arrays;

class Solution1790 {

    // T: O(n), S: O(1)
    public boolean areAlmostEqual(String s1, String s2) {
        var counts1 = new int[26];
        var counts2 = new int[26];
        int diffs = 0;
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                diffs++;
            }
            counts1[s1.charAt(i) - 'a']++;
            counts2[s2.charAt(i) - 'a']++;
        }
        if (!Arrays.equals(counts1, counts2)) {
            return false;
        }
        return diffs == 2 || diffs == 0;
    }

    public static void main(String[] args) {
        var sol = new Solution1790();
        System.out.println(sol.areAlmostEqual("bank", "kanb")); // true
        System.out.println(sol.areAlmostEqual("attack", "defend")); // false
        System.out.println(sol.areAlmostEqual("kelb", "kelb")); // true
        System.out.println(sol.areAlmostEqual("caa", "aaz")); // false
    }

}
